File: scripts/paths.py
"""
paths.py — every filesystem location the pipeline uses, resolved once.

WHY THIS EXISTS
    Each script used to carry its own copy of
        ROOT = os.environ.get("VIV_ROOT", "/Users/bjameshaskins/Desktop/_PROJECTS")
    and build paths from it. Six copies of one fact, a personal home directory baked
    into a shared repo, and a pipeline that only worked if you ran it from exactly the
    right directory. Locations are derived from this file's own position instead, so
    the working directory stops mattering.

LAYOUT
    This repo owns everything from photographs to seeded records:

        <repo>/scripts/           this file, and the pipeline
        <repo>/data/input/        drop zone
        <repo>/data/books/        grouped book folders (the queue)
        <repo>/data/archive/      shipped batches
        <repo>/data/intake/       staging copy, scratch
        <repo>/data/ready/        webp derivatives, scratch
        <repo>/data/_work/        ledger, grouping proposal, thumbnails

    The app owns the catalogue itself, and this pipeline writes two things into it:

        <app>/data/items.json     the records
        <app>/public/items/       the webp that get uploaded to R2

    Taxonomy, owner and location are the app's alone — nothing here reads or writes
    them, which is why the vocabulary file does not appear above.

    The app is assumed to be a sibling directory named `vivarium`. Override with
    VIVARIUM_APP if it lives elsewhere.
"""

import os

HERE = os.path.dirname(os.path.abspath(__file__))
REPO = os.path.dirname(HERE)

DATA = os.path.join(REPO, "data")
INPUT = os.path.join(DATA, "input")
BOOKS = os.path.join(DATA, "books")
ARCHIVE = os.path.join(DATA, "archive")
INTAKE = os.path.join(DATA, "intake")
READY = os.path.join(DATA, "ready")
WORK = os.path.join(DATA, "_work")

RECORDS_SOURCE = os.path.join(DATA, "records_source.json")

APP = os.environ.get("VIVARIUM_APP") or os.path.join(os.path.dirname(REPO), "vivarium")
ITEMS = os.path.join(APP, "data", "items.json")
PUBLIC_ITEMS = os.path.join(APP, "public", "items")

# The service-role key and R2 credentials stay in the app, in one copy. The node
# steps are invoked with --env-file pointing here rather than duplicating secrets.
ENV_FILE = os.path.join(APP, ".env.vivarium")


def require_app():
    """
    Fail early and legibly when the catalogue isn't where we think it is. Every write
    step depends on it, so finding out at step five is worse than finding out at step
    zero.
    """
    if not os.path.isfile(ITEMS):
        raise SystemExit(
            f"Cannot find the Vivarium app.\n"
            f"  expected: {ITEMS}\n"
            f"  set VIVARIUM_APP to the app directory if it lives elsewhere."
        )

File: scripts/paths.mjs
// paths.mjs — the node half of scripts/paths.py. Same locations, same rules.
//
// Resolved from this file's own position rather than process.cwd(), so it stops
// mattering which directory you run from. That used to matter a great deal: these
// scripts read `public/` and `data/items.json` relative to the working directory and
// only worked from the app root.
//
// The app is assumed to be a sibling directory named `vivarium`. Override with
// VIVARIUM_APP.

import path from 'node:path';
import { existsSync } from 'node:fs';
import { fileURLToPath } from 'node:url';

export const HERE = path.dirname(fileURLToPath(import.meta.url));
export const REPO = path.dirname(HERE);

export const DATA = path.join(REPO, 'data');
export const INTAKE = path.join(DATA, 'intake');
export const READY = path.join(DATA, 'ready');
export const RECORDS_SOURCE = path.join(DATA, 'records_source.json');

export const APP =
  process.env.VIVARIUM_APP || path.join(path.dirname(REPO), 'vivarium');

/**
 * Was the app directory named, or guessed?
 *
 * The default is a sibling called `vivarium`, which was correct when one
 * checkout served every instance. It is not correct now: each instance has its
 * own clone (vivarium, vivarium-tamplin, vivarium-sirsinate), and passing
 * --env-file changes which DATABASE a script reads without changing which TREE
 * it writes to. On 10 September 2026 a sync pointed at the Tamplin database
 * aimed 221 paintings at the library's tree; only the id-ahead guard in
 * sync_from_supabase.mjs stopped it.
 *
 * Callers that WRITE into the app should require this to be true. Reading is
 * harmless either way.
 */
export const APP_EXPLICIT = Boolean(process.env.VIVARIUM_APP);
export const APP_DATA = path.join(APP, 'data');
export const ITEMS = path.join(APP_DATA, 'items.json');
export const PUBLIC = path.join(APP, 'public');
export const PUBLIC_ITEMS = path.join(PUBLIC, 'items');

// Every write step depends on the app being where we think it is. Finding out at
// step zero beats finding out after a Supabase pull.
export function requireApp() {
  if (!existsSync(ITEMS)) {
    console.error(
      `Cannot find the Vivarium app.\n` +
        `  expected: ${ITEMS}\n` +
        `  set VIVARIUM_APP to the app directory if it lives elsewhere.`,
    );
    process.exit(1);
  }
}

/**
 * Refuse to write into a guessed tree when the environment names a specific
 * database.
 *
 * `--env-file` selects the database. VIVARIUM_APP selects the tree. Nothing
 * connects them, so the dangerous combination is an env file for one instance
 * and a defaulted app path pointing at another's clone. Naming the tree makes
 * the pairing deliberate.
 *
 * This does not catch every mismatch — a wrong VIVARIUM_APP is still wrong, and
 * no check here can know which database an env file belongs to. It catches the
 * one that actually happened.
 */
export function requireExplicitApp(what = 'write') {
  if (APP_EXPLICIT) return;
  console.error(
    `REFUSING TO ${what.toUpperCase()}: VIVARIUM_APP is not set, so the app directory was\n` +
      `guessed as:\n  ${APP}\n\n` +
      `--env-file chooses the DATABASE; VIVARIUM_APP chooses the TREE. With one\n` +
      `clone per instance they have to be named together, or a pull from one\n` +
      `instance lands in another's working copy.\n\n` +
      `Set it in that clone's .env.local, e.g.\n` +
      `  VIVARIUM_APP=/Users/you/Desktop/_PROJECTS/vivarium-tamplin\n` +
      `or pass it inline:\n` +
      `  VIVARIUM_APP=../vivarium-tamplin node --env-file=... scripts/<script>.mjs`,
  );
  process.exit(1);
}
